enterprise/frontend/src/metabase-enterprise/audit_app/containers/AuditTable.jsx

```diff
@@ -39,97 +26,60 @@ const mapDispatchToProps = {
 
 const DEFAULT_PAGE_SIZE = 100;
 
-@connect(
-  mapStateToProps,
-  mapDispatchToProps,
-)
-export default class AuditTable extends React.Component {
-  props: Props;
-  state: State = {
-    page: 0,
-    hasMorePages: false,
-  };
-
-  static defaultProps = {
-    pageSize: DEFAULT_PAGE_SIZE,
-  };
-
-  render() {
-    const { metadata, table, onChangeLocation, pageSize } = this.props;
-    const { page, hasMorePages } = this.state;
-
-    const card = chain(table.card)
-      .assoc("display", "audit-table")
-      .assocIn(["dataset_query", "limit"], pageSize)
-      .assocIn(["dataset_query", "offset"], pageSize * page)
-      .value();
+AuditTable.defaultProps = {
+  pageSize: DEFAULT_PAGE_SIZE,
+};
 
-    const question = new Question(card, metadata);
+AuditTable.propTypes = {
+  metadata: PropTypes.object.isRequired,
+  table: PropTypes.object.isRequired,
+  onChangeLocation: PropTypes.func.isRequired,
+  pageSize: PropTypes.number.isRequired,
+};
 
-    return (
-      <div>
-        <QuestionLoadAndDisplay
-          className="mt3"
-          question={question}
-          metadata={metadata}
-          mode={AuditMode}
-          onChangeLocation={onChangeLocation}
-          onChangeCardAndRun={() => {}}
-          onLoad={results =>
-            this.setState({ hasMorePages: results[0].row_count === pageSize })
-          }
-        />
-        {(hasMorePages || page > 0) && (
-          <div className="mt1 pt2 border-top flex">
-            <PaginationControls
-              className="ml-auto"
-              start={page * pageSize}
-              end={(page + 1) * pageSize - 1}
-              hasPrevious={page > 0}
-              hasNext={hasMorePages}
-              onPrevious={() => this.setState({ page: page - 1 })}
-              onNext={() => this.setState({ page: page + 1 })}
-            />
-          </div>
+function AuditTable({ metadata, table, onChangeLocation, pageSize, ...rest }) {
+  const [loadedCount, setLoadedCount] = useState(0);
+  const { handleNextPage, handlePreviousPage, page } = usePagination();
+
+  const card = chain(table.card)
+    .assoc("display", "audit-table")
+    .assocIn(["dataset_query", "limit"], pageSize)
+    .assocIn(["dataset_query", "offset"], pageSize * page)
+    .value();
+
+  const question = new Question(card, metadata);
+  const shouldShowPagination = page > 0 || loadedCount === pageSize;
+
+  return (
+    <div>
+      <QuestionLoadAndDisplay
+        className="mt3"
+        question={question}
+        metadata={metadata}
+        mode={AuditMode}
+        onChangeLocation={onChangeLocation}
+        onChangeCardAndRun={() => {}}
```

Comment:
Just curious, what's the point of passing an empty function here?

Reply:
Without passing `onChangeCardAndRun` click behavior does not work which makes "Name" column is not clickable. I decided not to invest in refactoring here because it is a part of visualization code that requires more investigation and extensive testing